bundles/org.openhab.binding.insteon/src/main/java/org/openhab/binding/insteon/internal/InsteonBinding.java

```diff
@@ -0,0 +1,517 @@
+ *
+ * This class represents the actual implementation of the binding, and controls the high level flow
+ * of messages to and from the InsteonModem.
+ *
+ * Writing this binding has been an odyssey through the quirks of the Insteon protocol
+ * and Insteon devices. A substantial redesign was necessary at some point along the way.
+ * Here are some of the hard learned lessons that should be considered by anyone who wants
+ * to re-architect the binding:
+ *
+ * 1) The entries of the link database of the modem are not reliable. The category/subcategory entries in
+ * particular have junk data. Forget about using the modem database to generate a list of devices.
+ * The database should only be used to verify that a device has been linked.
+ *
+ * 2) Querying devices for their product information does not work either. First of all, battery operated devices
+ * (and there are a lot of those) have their radio switched off, and may generally not respond to product
+ * queries. Even main stream hardwired devices sold presently (like the 2477s switch and the 2477d dimmer)
+ * don't even have a product ID. Although supposedly part of the Insteon protocol, we have yet to
+ * encounter a device that would cough up a product id when queried, even among very recent devices. They
+ * simply return zeros as product id. Lesson: forget about querying devices to generate a device list.
+ *
+ * 3) Polling is a thorny issue: too much traffic on the network, and messages will be dropped left and right,
+ * and not just the poll related ones, but others as well. In particular sending back-to-back messages
+ * seemed to result in the second message simply never getting sent, without flow control back pressure
+ * (NACK) from the modem. For now the work-around is to space out the messages upon sending, and
+ * in general poll as infrequently as acceptable.
+ *
+ * 4) Instantiating and tracking devices when reported by the modem (either from the database, or when
+ * messages are received) leads to complicated state management because there is no guarantee at what
+ * point (if at all) the binding configuration will be available. It gets even more difficult when
+ * items are created, destroyed, and modified while the binding runs.
+ *
+ * For the above reasons, devices are only instantiated when they are referenced by binding information.
+ * As nice as it would be to discover devices and their properties dynamically, we have abandoned that
+ * path because it had led to a complicated and fragile system which due to the technical limitations
+ * above was inherently squirrely.
+ *
+ *
+ * @author Bernd Pfrommer - Initial contribution
+ * @author Daniel Pfrommer - OpenHAB 1 insteonplm binding
+ * @author Rob Nielsen - Port to OpenHAB 2 insteon binding
+ */
+@NonNullByDefault
+public class InsteonBinding {
+    private final Logger logger = LoggerFactory.getLogger(InsteonBinding.class);
+
+    private Driver m_driver = new Driver();
```

Comment:
A general comment. A lot of fields use the convention `<character>_`. This is not very common.

Reply:
Yes I agree. That is the convention that the original developers used, m_ indicates member.

Reply:
`<character>_` has been refactored and doesn't exist anymore